cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -178,44 +220,40 @@ public String call(DatasetContext context) throws Exception {
     }
 
     /**
-     * Fetches and decodes new notifications and processes them
+     * Processes a list of notifications
      *
      * @param context the dataset context
-     * @param fetcher the message fetcher
-     * @return the last fetched message id
+     * @param notifications the list of fetched notifications from TMS
      * @throws Exception
      */
-    String fetchAndProcessNotifications(DatasetContext context, MessageFetcher fetcher) throws Exception {
-      String lastFetchedMessageId = null;
-      try (CloseableIterator<Message> iterator = fetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
-                                                               topic, 100, messageId)) {
-        LOG.trace("Fetch with messageId = {}", messageId);
-        while (iterator.hasNext() && !stopping) {
-          Message message = iterator.next();
-          Notification notification;
-          try {
-            notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
-                                         Notification.class);
-          } catch (JsonSyntaxException e) {
-            LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
-            lastFetchedMessageId = message.getId(); // update messageId to skip this message in next fetch
-            continue;
-          }
-          processNotification(context, notification);
-          lastFetchedMessageId = message.getId();
-        }
-
-        // Persist the last fetched message id
-        if (lastFetchedMessageId != null) {
-          updateMessageId(lastFetchedMessageId);
-        }
-
-        return lastFetchedMessageId;
+    public void fetchAndProcessNotifications(DatasetContext context, List<Notification> notifications)
+      throws IOException, DatasetManagementException {
+      for (Notification notification : notifications) {
+        processNotification(context, notification);
       }
     }
 
     protected JobQueueDataset getJobQueue() {
-      return Schedulers.getJobQueue(multiThreadDatasetCache, datasetFramework);
+      return Transactions.executeUnchecked(transactional, new TxCallable<JobQueueDataset>() {
```

Comment:
Actually, I guess every time I get the job queue, I can pass it the context, and use that to retrieve it every time. I will refactor to do that instead.